feat(hmr): adding hot module reloading

ScriptedAlchemy opened the pull request:
<!--
  HOLY CRAP a Pull Request. We ❤️ those!

  If you remove or skip this template, you'll make the 🐼 sad and the mighty god
  of Github will appear and pile-drive the close button from a great height
  while making animal noises.

  Please place an x (no spaces!) in all [ ] that apply
-->
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [x] **code refactor**
- [ ] **test update** <!-- if bug or feature is checked, this should be too -->
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

<!--
  Please explain the motivation or use-case for your change.
  What existing problem does the PR solve?
  If this PR addresses an issue, please link to the issue.
-->


Initial implementation copied over from extract-css-chunks-webpack-plugin

Adding Eslint config for loaders that are injected directly into the browser. 

These should remain ES5

### Breaking Changes

<!--
  If this PR introduces a breaking change, please describe the impact and a
  migration path for existing applications.
-->

### Additional Info
This is a work in progress, looking for feedback on how to implement the HMR part in a better way.
Once the is discussed, ill write the readme and commit it to this branch

This is the unification process of my work on [extract-css-chunks](https://github.com/faceyspacey/extract-css-chunks-webpack-plugin)

ScriptedAlchemy commented on src/index.js:
```diff
@@ -15,6 +19,24 @@ const REGEXP_CHUNKHASH = /\[chunkhash(?::(\d+))?\]/i;
+    }
+  }
+  return false;
+};
```
Yay! Thanks for all this feedback. I’ll ask some questions tomorrow. But completely agree with you. Thanks for reviewing this thoroughly

ScriptedAlchemy commented on src/hmr/.eslintrc.js:
```diff
@@ -0,0 +1,23 @@
+    'prefer-spread': 0,
+    'prefer-destructuring': 0
+  },
+};
```
okay, moving the code into the main build should solve the es5 issue. silly that I haven't done this already haha

ScriptedAlchemy replied:
@evilebottnawi how does one make a test for this?

ScriptedAlchemy commented:
Still need to test this further, but have been able to simplify it to pass most options via loader options. Testing filename output..

evilebottnawi commented:
@ScriptedAlchemy you can looks how we can tests loader in other repositories (example in `css-loader`)

ScriptedAlchemy commented:
@evilebottnawi I've gotten many of the builds to pass and updated the tests that need to be ( i will still add a HMR test case )

There seems to be an issue with node6

```TypeError: Cannot read property 'tap' of undefined```

ScriptedAlchemy commented:
@evilebottnawi im getting an error on canery

```TypeError: Cannot read property 'tap' of undefined``` 

Got any ideas? its passing everything else

ScriptedAlchemy commented on src/hmr/hotLoader.js:
```diff
@@ -0,0 +1,33 @@
+  `;
+}
+
+module.exports = hotLoader;
```
Took some thinking re-thinking, but its a far smoother implementation now! thanks

ScriptedAlchemy commented:
@evilebottnawi, ive gotten it to pass almost every test, but it seems to fail at `node6-canary` any ideas as to what is causing this??

ScriptedAlchemy commented on package.json:
```diff
@@ -43,8 +43,10 @@
   },
   "dependencies": {
-    "schema-utils": "^1.0.0",
+    "es-check": "^5.0.0",
```
ohh my bad!

ScriptedAlchemy replied:
moved it to devDependencies

evilebottnawi replied:
:+1:

ScriptedAlchemy commented:
> @ScriptedAlchemy don't see on `canary` tests (it is for `webpack@5`), plugin has no compatibility with `webpack@5` right now

oh wow, is Tappable and its hooks changing again

ScriptedAlchemy commented:
I will work on updating documentation this week

evilebottnawi commented:
@ScriptedAlchemy ping me when work was completed

ScriptedAlchemy commented:
@evilebottnawi this has been completed and I have released the codebase of this PR under my original plugin too. As this was a large refactor (and stability improvement) I wanted to take advantage of my existing user-base. https://www.npmjs.com/package/extract-css-chunks-webpack-plugin

I released it today as the latest version, so if there are any issues - my dependents will also report them via my plugin. This is likely to be the last release of extract-css-chunks as I focus on maintaining mini-css and will be pointing my dependents to this project moving forwards.

evilebottnawi commented:
Good job :+1:

ScriptedAlchemy commented:
One of my dependents has identified a small bug. Ill be patching it this weekend and will update this PR as well. Ill drop another comment once I have updated this PR

evilebottnawi commented on package.json:
```diff
@@ -74,7 +74,7 @@
     "prettier": "^1.11.1",
     "standard-version": "^4.3.0",
-    "webpack": "^4.28.0",
+    "webpack": "4.28.0",
```
Why?

ScriptedAlchemy replied:
Seemingly upgraded to webpack 4.29 and caused all tests with the import() syntax to fail. Locking it at least makes stuff pass. Not specifically a webpack problem, but incredibly irritating. Can try tinkering with it again

ScriptedAlchemy replied:
Finally got the green check ✅ with this though. I’ll check on the status of the npm-cli release. Hoisting supposed to be patched in its next release.

evilebottnawi commented:
/cc @ScriptedAlchemy All done? Can we review?

ScriptedAlchemy commented:
I’ve got one more patch I will push tonight. Then we are ready to review, there has been a lot of testing on my plugin. The last set of fixes have made it very stable. 

I’ll comment when I have applied the final fix tonight. After that, I’ll switch to maintaining mini-css:)

evilebottnawi commented:
@ScriptedAlchemy glad to hear, feel free to ping me

ScriptedAlchemy commented:
@evilebottnawi I’ve pushed the last fix

evilebottnawi commented on README.md:
```diff
@@ -62,8 +59,9 @@ module.exports = {
-              publicPath: '../'
+              // by default it uses publicPath in webpackOptions.output
+              publicPath: '../',
+              hot: process.env.NODE_ENV === 'development'
```
Let's rename `hot` to `hmr` (we use same name for `style-loader` and we rename `hot` to `hmr` in `webpack-dev-server` in next major release)

ScriptedAlchemy replied:
okay

evilebottnawi commented on README.md:
```diff
@@ -114,6 +117,52 @@ module.exports = {
+              hot: process.env.NODE_ENV === 'development',
+              // modules enables hmr of css modules which
+              // need to be handled differently compared to typical css files
+              modules: true,
```
Can we check `locals` and enable this by default if `css-loader` return locals https://github.com/webpack-contrib/css-loader/blob/master/src/index.js#L326?

ScriptedAlchemy replied:
Ive had quite a few issues with this, but ill release something and see if anyone complains

evilebottnawi commented on README.md:
```diff
@@ -114,6 +117,52 @@ module.exports = {
+              // a fallback option has been implemented to force css files to be reloaded
+              // reloadAll will force all stylesheets to be reloaded when receiving an HMR update
+              // this option is helpful when dealing with css-modules as they are harder to HMR
+              reloadAll: true,
```
When it can be happens? I think we should add more documentation about this option. Maybe we can check this on client side and reload all if something broken/event not handled/or something other unexpected?

ScriptedAlchemy replied:
Yeah this isnt an ideal one, but sometimes reloading **just wont work** this pretty much ensures reloading will work. 

I will remove this from my project as well... but it does offer an effective fallback

evilebottnawi replied:
@ScriptedAlchemy maybe you can find issues/examples where it really doesn't work? We can add this option after release and feedback from users, maybe we just have a bug and we can fix it without `reloadAll` :+1:

shepherdwind replied:
The `reloadAll` option is for dealing with commons-chunk, eg [this example)(https://github.com/shepherdwind/css-hot-loader/blob/master/examples/commons-chunk/webpack.config.js#L24). Default the update logic is that

1. a.css required by a.js, and we have output a.js, then
2. when a.css update, a.js will receive update message, so the hmr module will reload a.js

But, a.css maybe extract to common.css, in this condition, we should reload all css file.

evilebottnawi commented on src/hmr/hotModuleReplacement.js:
```diff
@@ -0,0 +1,165 @@
+/* eslint prefer-arrow-callback: 0 */
+/* eslint prefer-template: 0 */
+
+var normalizeUrl = require('normalize-url');
```
Normalize@3 has `const` and other ES next features https://github.com/sindresorhus/normalize-url/blob/v3.3.0/index.js#L3, will be great transpile this, but we can setup this in other PR (just copy/paste code and rewrite this on es2015)

ScriptedAlchemy replied:
what PR, a pr to normalize or here? I do know that some users have had issues - i wouldn't want this to release and cause problems right away

ScriptedAlchemy replied:
will downgrade to v2

evilebottnawi commented on src/hmr/hotModuleReplacement.js:
```diff
@@ -0,0 +1,165 @@
+
+module.exports = function(moduleId, options) {
+  if (noDocument) {
+    return noop;
```
I think better do `console.log` and add message about unavailable document and what is it required for work

ScriptedAlchemy replied:
okay will add

evilebottnawi commented on src/index.js:
```diff
@@ -393,6 +394,21 @@ class MiniCssExtractPlugin {
     });
   }
 
+  traverseDepthFirst(root, visit) {
```
Can't find where we use this

evilebottnawi commented on test/TestCases.test.js:
```diff
@@ -1,5 +1,6 @@
 import fs from 'fs';
 import path from 'path';
+import { exec } from 'child_process';
```
Let's use `execa` for testing

ScriptedAlchemy commented:
@evilebottnawi I have updated everything as you request, I also released all the changes to my own plugin. My users will report back in a day or two 

I have not yet released the removal of reloadAll as I want to first make sure everything works. 

Once this PR is merged, will it be faster to PR future updates or fixes? I tend to work on mine to improve the stability and support users. For smaller changes, will it take less time?  The plugin has been at `0.5.0` for as long as I've been using it :p

evilebottnawi commented:
@ScriptedAlchemy 
> Once this PR is merged, will it be faster to PR future updates or fixes? I tend to work on mine to improve the stability and support users. For smaller changes, will it take less time? The plugin has been at 0.5.0 for as long as I've been using it :p

Yes, it will be pretty fast, here is a big update so we can not do quickly, also as you can see we find some bugs and misleading.

Anyway good work!

ScriptedAlchemy commented:
Yeah it’s definitely a big feature and I’m glad I have a testing ground. I fully understand we must be careful:) just thinking about patches and fixes. 

Thanks for your input!

evilebottnawi commented:
@ScriptedAlchemy I think we are ready to merge, all works are done?

ScriptedAlchemy commented:
Let me latest release get some users, I want to make sure everything works. I’ve also asked some dependents to test the latest release. I might be able to let you know today otherwise tomorrow morning.

evilebottnawi commented:
Ping me when it is ready for merge :+1:

ScriptedAlchemy commented:
I need to test something after the last round of updates some of my users report occasional full page reloads. This week should be the week it’s merged

ScriptedAlchemy commented:
Okay there’s definitely something going on with css modules. Unable to HMR without the reload all option. I’m going to see if I can find the problem or at least automate the reloading.

ScriptedAlchemy commented:
@evilebottnawi I need some assistance. 

I found the cause of CSS modules not reloading. the module.id = `/src/containers/Home/style.scss`
but the actual file name is `containers-Home.css`

the best ive been able to do is

```     for (const module of modules) {
          for (const modChunks of Array.from(module._modules)) {
            if (modChunks._source && modChunks._source._value) {
              const newSource = modChunks._source._value.replace(
                /__pending_file__/g,
                module.id
              );
              modChunks._source._value = newSource;
            }
          }
        }```

however, this causes a syntax error and somehow breaks the hot loading script `_webpack_require(` becomes `r_webpack_require"`

for the time being, it might be best to use the reloadAll option of css modules...

I can prepare an example of this issue if you'd like to take a look

ScriptedAlchemy commented:
Heres the function i was trying to update, its injected by the loader

```function hotLoader(content, context) {
  const accept = context.locals
    ? ''
    : 'module.hot.accept(undefined, cssReload);';
  const result = `${content}
    if(module.hot) {
      // ${Date.now()}    
      console.log('__pending_file__');
      var cssReload = require(${loaderUtils.stringifyRequest(
        context.context,
        path.join(__dirname, 'hmr/hotModuleReplacement.js')
      )})(module.id, ${JSON.stringify(context.query)});
      module.hot.dispose(cssReload);
      ${accept}
    }
  `;
  return result;
}
```

ScriptedAlchemy commented:
I am able to automate `reloadAll` by checking if its locals or not. It would allow for a stable release....

evilebottnawi commented:
@ScriptedAlchemy :+1: Let's improve documentation about `reloadAll` and css modules and add tests/manual tests for this option

ScriptedAlchemy commented:
Awesome! I’ll start doing this today. Also going to fix the manual test (little webpack build in the repo) it’s currently broken and would be nice to have a playground for future development. 

Very excited to see this go live

ScriptedAlchemy commented:
I've added back the `reloadAll` and also automated it. If `locals` then it will reloadAll automatically. 

I also added an override to disable order warnings, this makes the terminal very messy. with this feature added in - mini-css is an exact copy of extract-css. I'll be able to keep them in sync and easily test new features on extract-css before rolling out.

I've done one last release on extract css to test the changes, but I think we could be ready to merge. If anyone would like to test this, it would be appreciated. I'll be watching the issues and will handle anything if there's a problem.

evenstensberg commented:
I can give it a spin at a large codebase and I’ll give you feedback on things, if, something shows up

evilebottnawi commented:
@evenstensberg wait you manual testing and we can merge and release this

@ScriptedAlchemy Thanks for great work!

evilebottnawi commented on README.md:
```diff
@@ -138,7 +186,9 @@ module.exports = {
-      chunkFilename: "[id].css"
+      chunkFilename: "[id].css",
+      // disables order warnings being logged in terminal after build
+      orderWarning: true, // Disable to remove warnings about conflicting order between imports
```
Let's avoid this option here, let's do it in other PR and add documentation how remove this using `warningsFilter`

ScriptedAlchemy commented:
Okay no problem. I’m also going to implement the updated code into my platforms at work. 

Will remove the order warning stuff and open it in another PR

evilebottnawi commented:
@ScriptedAlchemy :+1:

tehOPEologist commented:
just ran these changes against a project i'm working on successfully!  with these changes i'll be able to remove my css-hot-loader dependency, as it seems to do the same thing.  thanks so much for all your work!  excited for the release ^_^

ScriptedAlchemy commented:
Okay, we are ready to merge. I've removed order warning from this PR and will open it in another. @evilebottnawi take one last look and let me know.

evenstensberg commented:
Also we had some issues with babel v6 and v7 w.r.t polyfills and `preset-env`. Any idea where that might come from? mini-extract or webpack-dev-server? Is that issue on user or webpack?

ScriptedAlchemy commented:
Hmmmm that’s interesting, I’ve seen webpack dev server reload pages but usually it’s due to configuration error or not having a hot loader or something enabled. Any chance you’ve got a demo repo of this? 

Are you using normal css, sass, less, or css modules? 

I’ll definitely fix your issue if it’s in the plugin or webpack.

evilebottnawi commented:
@evenstensberg i think problem in configuration, can you provide minimum reproducible test repo?

dagda1 commented:
I've tested 4.3 of extract-css-chunk-webpack-plugin with both a static build and an ssr build and both had hmr working great after having problems with 4.2.

evenstensberg commented:
I think the repro is an edge case, no big deal

evilebottnawi commented:
So we ready to merge and release?

ScriptedAlchemy commented:
I’d say so, if anyone has a problem, I’m working from home today and will be on alert as well as the weekend I’ll be coding.

evilebottnawi commented:
@ScriptedAlchemy we merge and do release tomorrow, weekend for tests (in wild) and new issue(s) and we can start fix it since Monday

evilebottnawi commented:
Also i want fix CI after merge to avoid problem for other developers

ScriptedAlchemy commented:
Yeah me too! I’ll slack you on webpack channel? Want to figure   What the problem is

evilebottnawi commented:
@ScriptedAlchemy yep, feel free to write :+1:

ScriptedAlchemy commented:
@evilebottnawi merge this weekend still? 

Want to have some time before work to see if any issues show up. Harder to roll fixes at the office

evilebottnawi commented:
@ScriptedAlchemy sorry for delay on weekend, i was ill, we merge this today and release, bugs is bad but we can avoid their :smile: So we can release this and get feedback from developers and fix their in future

evilebottnawi commented:
/cc @ScriptedAlchemy need rebase :+1:

ScriptedAlchemy commented:
OKay will do

ScriptedAlchemy commented:
@evilebottnawi I've updated the branch

evilebottnawi commented:
@ScriptedAlchemy Thanks

evilebottnawi commented:
Today i fix CI problem, update deps and boilerplate for plugin and do release

tehOPEologist commented:
when can we expect a version bump + npm publish for this feature?  just waiting on that to migrate our implementation to this!  thanks again

ScriptedAlchemy commented:
There’s a few things being finalized in another branch. Mostly chores needed for the project

zackschuster commented:
is there a reason to use `normalize-url@^2.0.1`? i ask because it'd be nice to have `normalize-url^3.0.0` to line up with `postcss-normalize-url@~4.0.0` (aka "what `optimize-css-assets-webpack-plugin` is doing")

thanks for your hard work! :smile:

evilebottnawi commented:
`normalize-url` use modern ES syntax so we need setup babel for transpile modern ES syntax to ES5 for client, you can send a PR

evilebottnawi commented:
/cc @ScriptedAlchemy we should solve this in next PR

zackschuster commented:
@evilebottnawi thanks for the explanation, i'll try to PR if i have the time!

ScriptedAlchemy commented:
It uses ES6 stuff that causes problem on older node versions
